### ISSUE TITLE
Image size is changed when do model.predict

### ISSUE THREAD

**Issue by iagupov** [ultralytics/ultralytics#3955]
### Search before asking

- [x] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, could you please provide me with some clarification? 
I trained a YOLOv8n model on fullHD resolution with imgsz=[1920, 1080]. After that I did model.predict on an images with regular full HD resolution of 1920x1080. The parameters: imgsz=[1920, 1080], rect=True. As expected, my image was resized to 1920x1088, which is nothing unusual. However, when the model started to make image-by-image inference, the resolution changed to 640x1088. Please find logs below:
```
WARNING  imgsz=[1920, 1080] must be multiple of max stride 32, updating to [1920, 1088]
image 1/5 C:\Users\user\images\frame_0072.png: 640x1088 3 cars, 226.4ms
image 2/5 C:\Users\user\images\frame_0090.png: 640x1088 5 cars, 193.5ms
image 3/5 C:\Users\user\images\frame_0108.png: 640x1088 2 cars, 174.6ms
image 4/5 C:\Users\user\images\frame_0135.png: 640x1088 10 cars, 215.4ms
image 5/5 C:\Users\user\images\frame_0180.png: 640x1088 6 cars, 182.5ms
Speed: 8.8ms preprocess, 198.5ms inference, 1.0ms postprocess per image at shape (1, 3, 1920, 1088)
```
and whats I find strange is that when I inverted imgsz to be [1080, 1920] (while using the same images with a normal resolution of 1920x1080), the image resolution became as it should be, but inverted. And an inference time became about 3 times longer (I suspect that in this case yolo really makes an inference on a fulll hd image). Please find longs below:

```
WARNING  imgsz=[1080, 1920] must be multiple of max stride 32, updating to [1088, 1920]
image 1/5 C:\Users\user\images\frame_0072.png: 1088x1920 2 cars, 672.2ms
image 2/5 C:\Users\user\images\frame_0090.png: 1088x1920 3 cars, 622.3ms
image 3/5 C:\Users\user\images\frame_0108.png: 1088x1920 3 cars, 614.3ms
image 4/5 C:\Users\user\images\frame_0135.png: 1088x1920 5 cars, 621.3ms
image 5/5 C:\Users\user\images\frame_0180.png: 1088x1920 9 cars, 640.3ms
Speed: 18.5ms preprocess, 634.1ms inference, 1.0ms postprocess per image at shape (1, 3, 1088, 1920)
```

It would be very helpful if anyone could explain this behaviour because I didn't find any explanation or at least a similar problem.

Thanks in advance for your assistance,
Best regards,
Ilia

**Comment by glenn-jocher** [comment]
@iagupov hello Ilia,

Thank you for bringing this issue to us. The YOLOv8 structure that we have developed does require that the input image resolution be a multiple of the maximum network stride, here being 32, to ensure the images are processed correctly through the layers of the model. As such, the input dimensions are adjusted if necessary at the preprocessing stage, which explains the updated image dimensions you have noticed in the logs.

The reduction in resolution on inference to 640x1088 is a product of the YOLOv8 model's design to ensure efficient use of GPU memory and faster runtimes. The aspect ratio of the input image is preserved, and the longer dimension of the image is downsized to an upper limit (in this case 640 pixels) at the expense of the resolution.

When you switch the imgsz parameter to [1080, 1920], the model takes the first value to be the short axis (height) of the image. Consequently, the larger dimension (width) of the image is taken as the height, leading to a larger total resolution for the processed image and a longer inference time due to the additional computations involved with the larger input size.

I hope this provides some clarity. Further theoretical details on this subject are described across various YOLOv8-related papers and documentation. Let us know if you have more questions.

Best regards,
Glenn.

**Comment by iagupov** [comment]
@glenn-jocher Thank you so much for your quick and clear reply! I would like to ask one more question. I chose 1920x1080 resolution because I want to detect small objects, and I'm afraid that reducing it from 1920 to 640 pixels may harm the final precision. Maybe there is a flag that I could control to prevent this reducton?

Thank you!
Best regards,
Ilia

**Comment by glenn-jocher** [comment]
@iagupov Hello Ilia,

You're correct that the reduction in resolution during inference may potentially impact the detection of smaller objects. To address this, YOLOv8 provides a flexibility where you can adjust the inference image size to better suit your particular use case. 

In your situation, to prevent the reduction of image resolution, you might want to adjust the --img-size argument used in your inference command. By increasing the --img-size value, you'll be able to perform inference at a higher resolution, and this could help improve the detection of your small objects.

Remember that increasing the inference image size will also increase GPU memory usage and potentially the processing time. Therefore, please ensure that your GPU has enough memory to handle the increased image size.

I hope this helps and let us know if you have further questions!

Best regards,
Glenn

**Comment by iagupov** [comment]
@glenn-jocher Hello Glenn. Am I correct if I say that during inference the variable `imgsz` = `[high, width]`? I precise at this point because in the documentation it says opposite that _width_ is on the first place and _high_ on the second place.
> imgsz: 640  # (int) size of input images as integer or w,h

And if I want to do an inference on a regular HD image having width=640 and high=480 I should do 
`model.predict(source=..., imgsz=[480, 640])`?

And moreover in this case I have the following log:
> Speed: 1.8ms preprocess, 83.0ms inference, 0.4ms postprocess per image at shape (1, 3, 480, 640)

where the shape corresponds to pytorch tensor notation of [batch, channels, high, width]

Best regards,
Ilia

**Comment by glenn-jocher** [comment]
@iagupov hello Ilia,

Great observation and question. You're correct that the PyTorch tensor uses the format [batch, channels, height, width], and this is what the shape notation in your log represents. Despite this, for the `imgsz` parameter in the `model.predict()` method, you still define it in the format as `[width, height]`. This might seem confusing initially but is maintained for consistency across the entire codebase.

So, if you're running inference on a regular HD image with dimensions 640 x 480, you would indeed use `model.predict(source=..., imgsz=[640, 480])`.

Hope this clarifies things for you!

Best Regards,
Glenn

**Comment by iagupov** [comment]
@glenn-jocher Hello Glenn, I'm sorry for questionning you again, but I still can not get a proper image output size. 
I trained a model with this parameters:
`model = YOLO('yolov8n.pt')`
`model.train(data='my_data.yaml', imgsz=[640, 480], rect=True)`

then I made an inference with following parameters:
`model.predict(source='images', imgsz=[640, 480], rect=True)`
and I got the following logs:

> image 1/5 C:\images\scene00072.png: **384x480** 1 car, 63.8ms
> image 2/5 C:\images\scene00084.png: **384x480** 5 cars, 48.9ms
> image 3/5 C:\images\scene00096.png: **384x480** 5 cars, 38.2ms
> image 4/5 C:\images\scene00108.png: **384x480** 9 cars, 46.9ms
> image 5/5 C:\images\scene00120.png: **384x480** 7 cars, 64.8ms
> Speed: 1.6ms preprocess, 52.5ms inference, 1.0ms postprocess per image at shape (1, 3, 640, 480)

then I changes `imgsz `to be = `[640, 640]` and my logs are:

> image 1/5 C:\images\scene00072.png: **480x640** 3 cars, 86.8ms
> image 2/5 C:\images\scene00084.png: **480x640** 6 cars, 91.8ms
> image 3/5 C:\images\scene00096.png: **480x640** 6 cars, 95.0ms
> image 4/5 C:\images\scene00108.png: **480x640** 5 cars, 81.8ms
> image 5/5 C:\images\scene00120.png: **480x640** 4 cars, 90.8ms
> Speed: 1.8ms preprocess, 89.2ms inference, 0.6ms postprocess per image at shape (1, 3, 640, 640)

then I chaned to `imgsz=[480, 640]` and I got:

> image 1/5 C:\images\scene00072.png: **480x640** 3 cars, 106.7ms
> image 2/5 C:\images\scene00084.png: **480x640** 6 cars, 96.7ms
> image 3/5 C:\images\scene00096.png: **480x640** 6 cars, 85.8ms
> image 4/5 C:\images\scene00108.png: **480x640** 5 cars, 94.7ms
> image 5/5 C:\images\scene00120.png: **480x640** 4 cars, 92.8ms
> Speed: 2.6ms preprocess, 95.3ms inference, 0.6ms postprocess per image at shape (1, 3, 480, 640)

and finally, just to check, I increased a width to 1280, so `imgsz=[1280, 480]`, so logs are:

> image 1/5 C:\images\scene00072.png: **384x480** 1 car, 69.8ms
> image 2/5 C:\images\scene00084.png: **384x480** 5 cars, 59.8ms
> image 3/5 C:\images\scene00096.png: **384x480** 5 cars, 57.5ms
> image 4/5 C:\images\scene00108.png: **384x480** 9 cars, 45.9ms
> image 5/5 C:\images\scene00120.png: **384x480** 7 cars, 42.9ms
> Speed: 2.0ms preprocess, 55.2ms inference, 1.0ms postprocess per image at shape (1, 3, 1280, 480)

And I also checked my images, that they are really 640x480 images where 640 is a width ↔ and 480 is a height ↕ dimention. Maybe I miss something very simple..but I can't get my 640x480 resolution during inference.

Thanks in advance for your assistance,
Best regards,
Ilia

**Comment by glenn-jocher** [comment]
@iagupov Hello Ilia,

Your confusion is understandable. Here's how it's working:

When you specify `imgsz=[640, 480]` and `rect=True`, YOLOv8 will aim to process images at 640x480 resolution while preserving the original aspect ratio and potentially creating empty spaces in the image.

However, because the image resolution needs to be a multiple of the network stride (which is 32 for YOLOv8), the resolution is automatically adjusted to the nearest lower multiple of 32 that also preserves the aspect ratio. This is why you see 384x480 as the final resolution in the logs instead of the inputted 640x480. 

Additionally, as mentioned in previous conversation, YOLOv8 uses the first dimension provided (`imgsz[0]`) to represent the longer side of the image, so when you specify `imgsz=[480, 640]` and `rect=True`, you effectively request a rectangular inference resolution of 640x480 in your landscape orientation images.

In conclusion, when using `rect=True`, the actual inference size is a function of both the specified `imgsz` and the original aspect ratio while satisfying the condition of being divisible by the network stride. 

I hope this clears up your confusion.

Best regards,
Glenn.

**Comment by iagupov** [comment]
@glenn-jocher Hello Glenn,

> Additionally, as mentioned in previous conversation, YOLOv8 uses the first dimension provided (imgsz[0]) to represent the longer side of the image, so when you specify imgsz=[480, 640] and rect=True, you effectively request a rectangular inference resolution of 640x480 in your landscape orientation images.

so would it be correct if I say that if I need to make an inference on my 640x480 image and if I want to keep resolution to be the original one (without reduction) I can indicate `imgsz=[480, 640]` ?
And I found also another approach to avoid image reduction, but I'm not really sure is it a good way? In a file augment.py/class LetterBox/ _ _ call _ _, I hard-coded scale ratio `r=1`.

To prevent the reduction of image resolution, previusly you adviced to increase an imgsz during inference, but unfortinately I didn't really get how to do it technically. Should I just increase it twice (for example) - and use `imgsz = [1280, 960]` (for example) instead of `imgsz = [640, 480]` during inference?

And one small question about my prevous message. So when I do an inference on landscape oriented image with 640x480 resolution and in logs I have this:

> image 1/5 C:\images\scene00072.png: 480x640 3 cars, 106.7ms

here the resolution it written in opposite way? Height (480) is first and width (640) is second?

Thank you very much for your replys,
Best regards,
Ilia

**Comment by glenn-jocher** [comment]
@iagupov Hello Ilia,

Regarding keeping the resolution of your 640x480 image the exact original size, yes, you are correct. In the `predict()` method, you can set the `imgsz` argument to [480, 640] to maintain this aspect ratio. Just remember that the first dimension you provide (`imgsz[0]`) corresponds to the longer side of your image.

Concerning your modification to set `r=1` in `Letterbox.__call__`, I wouldn’t recommend this approach because it could potentially cause other parts of the process to malfunction. The image scaling and padding operations have been carefully designed to ensure the best results.

When I suggested increasing the `imgsz` during inference, I meant that you could try using higher numbers. However, keep in mind that `imgsz` should correspond to your specific use case and the characteristics of your images, so doubling the values might not always be the best choice. It heavily depends on your specific scenario.

Finally, about the log output, you are correct. The specification "480x640" indeed represents height x width for the processed image during inference.

Hope this helps!

Best Regards,
Glenn.

**Comment by iagupov** [comment]
@glenn-jocher Hello Glenn, thank you so much, now it's entirely clear! And thank you for your patience, clear detailed answers and your guidance, your help is really great!
Best regards,
Ilia

**Comment by glenn-jocher** [comment]
@iagupov Hello Ilia, 

You're very welcome! I'm glad to hear that you found the explanations helpful. Don't worry about asking questions, that's how we all learn and improve. We are always here to help with your queries. Remember, no question is too small or too big. 

Thank you for your kind words!

Best regards,
Glenn.

**Comment by jwee1369** [comment]
Hi, I wanted to chime in on this:

> @iagupov hello Ilia,
> 
> Great observation and question. You're correct that the PyTorch tensor uses the format [batch, channels, height, width], and this is what the shape notation in your log represents. Despite this, for the `imgsz` parameter in the `model.predict()` method, you still define it in the format as `[width, height]`. This might seem confusing initially but is maintained for consistency across the entire codebase.
> 
> So, if you're running inference on a regular HD image with dimensions 640 x 480, you would indeed use `model.predict(source=..., imgsz=[640, 480])`.
> 
> Hope this clarifies things for you!
> 
> Best Regards, Glenn

From the above discussion, it's stated during predict to use imgsz=[width, height] ie. if the image is 640 wide and 480 high

However from the doc (https://docs.ultralytics.com/modes/predict/#inference-arguments) it's stated to use (height, width)

"image size as scalar or (h, w) list, i.e. (640, 480)"

From my runs it looks like it's following (h,w)

for example if predict(imgsz=[1920, 1080]) - assuming imgsz=[height, width] format - and my input image is 1920 wide x 1080 high
i get: 

`Speed: 4.5ms preprocess, 122.9ms inference, 1.3ms postprocess per image at shape (1, 3, 640, 1088)`

following torch format for CNN (batch, channel, height, width) this means:
1. width is 1088
      a. since we set imgsz width to 1080 it rounded to the next highest 32x which is 1088
      b. so our input image got shrunk from 1920 wide to 1088
2. height is 640
      a. this factor is dependent in 1. since 1920 width is the bigger values and it got shrunk to 1088 (1.765x). so 1080 / 1.765 -> 612 -> nearest multiple of 32x is 640

@glenn-jocher @iagupov 
based on your findings do you agree that during :
1. training: use imgsz=[width, height] per 
[https://docs.ultralytics.com/modes/train/#arguments](url)
3. predict, export use imgsz=[height, width] per [https://docs.ultralytics.com/modes/predict/#inference-arguments](url)

**Comment by glenn-jocher** [comment]
@jwee1369 hello,

Thank you for your analysis and inquiry!

It seems that there's a little confusion, I apologize for any lack of clarity. You are correct in noting that PyTorch tensors use the format [batch, channels, height, width]. However, the `imgsz` argument actually does not follow the PyTorch tensor orientation, it represents a kind of declaration of the longest side of your images since it is applied adaptively based on the original image aspect ratio.

In the `predict()` method (and other similar methods), `imgsz`'s first argument is used to represent the longest side of an image, regardless of whether the image is in portrait or landscape mode. Therefore, it could represent either width or height depending on the image's orientation.

So when you set `imgsz=[1920, 1080]`, YOLOv8 treats 1920 as the longest side. If an image is landscape (wider), 1920 represents the width, while if an image is portrait (taller), 1920 represents the height instead. Also, it's important to note the resolution needs to be a multiple of the network stride (which is 32 for YOLOv8), hence the rounding to the nearest lower multiple of 32.

I hope this response provides the clarity you're looking for. I've made note of the potential confusion, and we'll consider a more explicit explanation for this in our documentation in the future. Thanks for bringing up this interesting point!

Best regards,
Glenn.

**Comment by jwee1369** [comment]
Hi @glenn-jocher 

could you explain the following results all with input image size of `1920 wide x 1080 high`

1. `predict(imgsz=[800, 640])`, inference shape: `(1, 3, 384, 640)`
2. `predict(imgsz=[640, 800])`, inference shape: `(1, 3, 480, 800)`
3. `predict(imgsz=[800, 320])`, inference shape: `(1, 3, 192, 320)`
4. `predict(imgsz=[320, 800])`, inference shape: `(1, 3, 320, 576)`

if the first argument in the imgsz represent the long side of the image (1920) as you have stated, looking at the first example above, you would:
1. 1920 -> 800 = 2.4x, and to maintain aspect ratio, 1080 /2.4 (next highest 32x) = 480
2. this should give us inference shape of `(1, 3, 480, 800)` but instead the image was changed to `(1, 3, 384, 640)`

Also, could you steer me to the part of code that resizes the inference images? I've only managed to find the check_imgsz() in checks.py which seems to resize to the nearest 32x

Thanks

**Comment by glenn-jocher** [comment]
@jwee1369 hello,

When you use the `predict()` method with the `imgsz` parameter, it doesn't necessarily resize your image strictly according to the values you input. Instead of straightforwardly treating `imgsz=[width, height]` or `imgsz=[height, width]`, YOLOv8 treats `imgsz[0]` as the longer side of your image and `imgsz[1]` as the shorter side. It then adjusts your image while maintaining the original aspect ratio.

Looking at your first example where you used `predict(imgsz=[800, 640])` on a 1920x1080 image:

1.  `imgsz[0]` (800) isn’t the width but represents the longest side. So the target size for resizing will be 800 along the longer side, and the aspect ratio is kept intact. This is why the width becomes 640 (the nearest multiple of 32 less than 800 * 1080 / 1920 = 640).

2. 1920 -> 800 gives a resize factor of 2.4x. But to maintain aspect ratio, the height 1080 is divided not by 2.4 but by 1920 ÷ 800 = 2.4 (to maintain aspect ratio) = 450, which is rounded to the lower nearest 32 and becomes 448.

Therefore, the image gets resized to `(1, 3, 448, 800)`. The shape you see in logs `(1, 3, 384, 640)` represents the height, width after padding added by `Letterbox` transformation to make resolution multiple of 32.

As for the code that resizes inference images, it is in the file `datasets.py`, the class `Letterbox`. This class is responsible for resizing and the aspect-preserving letterbox operation.
  
I hope this helps clarify the behavior you're seeing!

Best Regards,
Glenn.

**Comment by chandra-ps612** [comment]
@glenn-jocher Hello,
I've had a similar confusion. Suppose I use ```imgsz=[416, 416]``` in ```predict()``` method. where, ```imgsz[0]``` and ```imgsz[1] represent longer and shorter side of an image respectively.
I've two sets of images that have a shape e.g. (4032, 3024) and (434, 650) respectively where ```img.shape[0]`` and ```img.shape[1]``` represent original height and weight respectively.
**Case I** 
1. After resizing the shape of torch tensor will be : (1, 3, 416, 315)
2. After applying ```Letterbox``` and padding the shape of torch tensor will be: (1, 3, 416, 288)
Note : Here, 288 is the nearest multiple of 32 less than 315 and resize factor is 9.6

**Case-II**
1. After resizing the shape of torch tensor will be : (1, 3, 289, 416)
2. After applying ```Letterbox``` and padding the shape of torch tensor will be: (1, 3, 288, 416)
Note : Here, 288 is the nearest multiple of 32 less than 289 and resize factor is 1.5
So, Am i doing it correctly?

**Comment by chandra-ps612** [comment]
@glenn-jocher Hello,
I've had a similar confusion. Suppose I use ```imgsz=[416, 416]``` in ```predict()``` method. where, ```imgsz[0]``` and ```imgsz[1]``` represent longer and shorter side of an image respectively.
I've two sets of images that have a shape e.g. (4032, 3024) and (434, 650) respectively where ```img.shape[0]```  and ```img.shape[1]``` represent original height and width respectively.
**Case-I** 
1. After resizing the shape of image tensor will be : (1, 3, 416, 312)
2. After applying padding the shape of image tensor will be: (1, 3, 416, 320)
Note : Here, resize factor is 9.692307692

**Case-II**
1. After resizing the shape of image tensor will be : (1, 3, 278, 416)
2. After applying padding the shape of image tensor will be: (1, 3, 288, 416)
Note : Here, resize factor is 1.5625

***PyTorch tensor notation : ([Batch, Channels, Height, Width])
So based on your last comment, Am i doing it correctly?

**Comment by glenn-jocher** [comment]
@chandra-ps612 hello,

Yes, you are on the right track with your understanding of the `Letterbox` transformation and the `predict()` method's `imgsz` parameter.

For **Case-I**, with an original image size of (4032, 3024), when you use `imgsz=[416, 416]`, the longer side of the image (4032) will be resized to 416, and the shorter side (3024) will be resized while maintaining the aspect ratio, resulting in a tensor shape of (1, 3, 416, 312) before padding. Padding is then applied to reach a width that is a multiple of the network stride, leading to a final tensor shape of (1, 3, 416, 320).

For **Case-II**, with an original image size of (434, 650), the longer side (650) will be resized to 416, and the shorter side (434) will be resized accordingly, giving you a tensor shape of (1, 3, 278, 416) before padding. Padding is added to make the height a multiple of the network stride, resulting in a final tensor shape of (1, 3, 288, 416).

Your calculations and understanding of the resize factor and the final tensor shapes are correct, following the PyTorch tensor notation ([Batch, Channels, Height, Width]).

**Comment by jwee1369** [comment]
Hi @glenn-jocher 

sorry for the long break

referring back to the above example of what you mentioned

> Looking at your first example where you used predict(imgsz=[800, 640]) on a 1920x1080 image:

>imgsz[0] (800) isn’t the width but represents the longest side. So the target size for resizing will be 800 along the longer side, and the aspect ratio is kept intact. This is why the width becomes 640 (the nearest multiple of 32 less than 800 * 1080 / 1920 = 640).

>1920 -> 800 gives a resize factor of 2.4x. But to maintain aspect ratio, the height 1080 is divided not by 2.4 but by 1920 ÷ 800 = 2.4 (to maintain aspect ratio) = 450, which is rounded to the lower nearest 32 and becomes 448.

>Therefore, the image gets resized to (1, 3, 448, 800). The shape you see in logs (1, 3, 384, 640) represents the height, width after padding added by Letterbox transformation to make resolution multiple of 32.


I am following your calculations

but why does the log output (1,3,384,640) instead of (1,3,448,800) using redict(imgsz=[800, 640]) on a 1920x1080 as you mentioned

Thanks for your help

**Comment by glenn-jocher** [comment]
@jwee1369 hello,

Apologies for any confusion. The log output (1, 3, 384, 640) indeed reflects the final tensor shape after the `Letterbox` transformation, which includes both resizing and padding. The resizing step scales the image to maintain the aspect ratio with the longest side being 800, and then padding is applied to ensure the dimensions are multiples of 32, which is a requirement for the network. The discrepancy in the expected output shape is likely due to a misunderstanding of the padding step that follows the resizing.

Regarding visualizing the resized images, the `visualize` flag typically overlays predictions on the original image. To visualize the resized and padded images, you would need to modify the code to access the intermediate tensor after the `Letterbox` transformation but before inference. This isn't directly supported by the current API and would require custom code changes.

I hope this clarifies your query.

**Comment by jwee1369** [comment]
Hi @glenn-jocher 

i dug a little deeper into the Letterbox class. It looks like the following code takes the new vs old height of either the width and height - whichever that is lowest
`r = min(new_shape[0] / shape[0], new_shape[1] / shape[1])`

1.  in the case of the example 1 above: min(800/1080, 640/1920) -> min(0.74, 0.33) -> 0.33
2. then we get unpad_size = (h,w) = (0.33 * 1080, 0.33 * 1920) = (360,640)
3. make sure they are divisible by 32 -> pad up -> (384, 640)

Correct me if i am wrong, when you say the first argument imgsz[0] assumes the longest, it looks like it isn't the case, since the code above compares imgsz[0] to the shape[0] which is the height. And the axis that determines the ratio is the lowest ratio of (new/old)

sorry for the long back and forth, this is all very new to me still

Appreciate you support

**Comment by glenn-jocher** [comment]
Hello @jwee1369,

You are correct in your understanding of the `Letterbox` class. The code snippet you've referenced indeed calculates the resize ratio `r` by taking the minimum of the new vs old aspect ratios for both width and height. This ensures that the resized image maintains the original aspect ratio without exceeding the specified `imgsz` dimensions.

Your step-by-step breakdown is accurate:
1. The minimum ratio is determined by comparing the ratios of the requested dimensions to the original image dimensions.
2. The original image is then resized using this ratio, resulting in the `unpad_size`.
3. Finally, the resized dimensions are adjusted to ensure they are divisible by the network stride, which is 32.

The `imgsz` parameter specifies the maximum allowed dimensions for the resized image, and the actual resizing is governed by the aspect ratio preservation logic you've described.

Thank you for your diligence in exploring the code, and no worries about the back and forth. I'm here to help, and I'm glad to see you're getting a solid grasp of how the resizing works!

**Comment by jwee1369** [comment]
Hi @glenn-jocher 

Thanks for the confirmation, i can now move forward with clarity :)

A few more related question:
1. if the inference step is scaling and padding image to final size from  w,h (800,640) to (384, 640) when the input is (1080, 1920), what size should i be training with? 
2. Also, i am getting the following warning when training with rect
     - WARNING ⚠️ 'rect=True' is incompatible with Multi-GPU training, setting 'rect=False'
     - WARNING ⚠️ updating to 'imgsz=800'. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer, i.e. 'yolo export imgsz=640,480' or 'yolo export imgsz=640'

   does training always pad it to square?

**Comment by glenn-jocher** [comment]
Hello @jwee1369,

1. For training, you should use an `imgsz` that closely matches the inference size you plan to use, but it can be larger to provide the network with more detail during training. If you're using `(384, 640)` for inference, you might consider training with `imgsz=640` or a similar value that's a multiple of 32.

2. The `rect` training option is designed for single-GPU training and allows for rectangular training batches, which can be more efficient. However, it's not compatible with Multi-GPU training, hence the warning. During training, if `rect=False`, images are padded to square shapes by default to maintain compatibility across different GPU architectures.

**Comment by jwee1369** [comment]
Hi @glenn-jocher 

If I understood correclty, the only reason you would choose rect=True during training and inference is to reduce RAM needed for each epoch and inference time, since padding is added to make it square

I have yet to test this and was wondering if you have first-hand data, but aside from speed, if i am planning to inference at (384, 640), would the following two different training approaches lead to about the same performance during training, and what if during inference, the opposite is used

1.  rect=True (384, 640) during training -> rect=False (640,640) during inference
2. rect=False (640,640) during training -> rect=True (384,640) during inference

Thanks!
Jason.

**Comment by glenn-jocher** [comment]
@jwee1369 hello Jason,

Yes, using `rect=True` can reduce memory usage and potentially speed up training and inference by avoiding unnecessary padding for datasets with non-square images.

Regarding your question, the performance of the model can vary depending on the training and inference settings. Generally, it's best to train with the same aspect ratio and resolution you plan to use during inference for optimal results. However, YOLO models are quite robust, and training with `rect=False` (square padding) and inferring with `rect=True` (rectangular) or vice versa can still yield good performance, although it might not be optimal.

It's recommended to experiment with both approaches to see which one works best for your specific use case and dataset.

**Comment by jwee1369** [comment]
Thanks for your input @glenn-jocher , will run some experiements, have a great weekend!

**Comment by glenn-jocher** [comment]
@jwee1369 you're welcome, Jason! Happy experimenting, and have a fantastic weekend! 😊👍

**Comment by Shichang-Ye** [comment]
Hi Glen,

Thank you for your information. Could you please let me know if there is a way for me to get the actual resized width and height of the image being processed?

Thanks.

**Comment by glenn-jocher** [comment]
Hello,

You're welcome! To get the actual resized width and height of the image being processed, you can access the `shape` attribute of the image tensor after it has been resized and padded. Here's a quick example using the `predict` method:

```python
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n.pt")

# Run inference on an image
results = model.predict("path/to/your/image.jpg")

# Access the resized image dimensions
for result in results:
    resized_shape = result.orig_shape  # Original shape before resizing
    processed_shape = result.img.shape  # Shape after resizing and padding
    print(f"Original shape: {resized_shape}, Processed shape: {processed_shape}")
```

This will give you the dimensions of the image after it has been resized and padded for inference. If you have any further questions or need additional assistance, feel free to ask!

**Comment by glenn-jocher** [comment]
Hello,

Thank you for your feedback and for pointing that out! It looks like I made an error in my previous response. The `Results` object does not have an `img` attribute. Instead, you can access the resized image dimensions directly from the model's preprocessing steps.

Here's an updated approach to get the resized dimensions:

```python
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n.pt")

# Run inference on an image
results = model.predict("path/to/your/image.jpg")

# Access the resized image dimensions
for result in results:
    original_shape = result.orig_shape  # Original shape before resizing
    processed_shape = result.imgs[0].shape  # Shape after resizing and padding
    print(f"Original shape: {original_shape}, Processed shape: {processed_shape}")
```

The `imgs` attribute within the `Results` object contains the processed images, and you can access their shapes as shown above.

If the issue persists or if you encounter any other problems, please ensure you are using the latest version of the Ultralytics package. Additionally, providing a minimum reproducible example can help us diagnose the issue more effectively. You can find more information on creating a reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any more questions or need further assistance! 😊

**Comment by markmno** [comment]
> Hello,
> 
> Thank you for your feedback and for pointing that out! It looks like I made an error in my previous response. The `Results` object does not have an `img` attribute. Instead, you can access the resized image dimensions directly from the model's preprocessing steps.
> 
> Here's an updated approach to get the resized dimensions:
> 
> ```python
> from ultralytics import YOLO
> 
> # Load a model
> model = YOLO("yolov8n.pt")
> 
> # Run inference on an image
> results = model.predict("path/to/your/image.jpg")
> 
> # Access the resized image dimensions
> for result in results:
>     original_shape = result.orig_shape  # Original shape before resizing
>     processed_shape = result.imgs[0].shape  # Shape after resizing and padding
>     print(f"Original shape: {original_shape}, Processed shape: {processed_shape}")
> ```
> 
> The `imgs` attribute within the `Results` object contains the processed images, and you can access their shapes as shown above.
> 
> If the issue persists or if you encounter any other problems, please ensure you are using the latest version of the Ultralytics package. Additionally, providing a minimum reproducible example can help us diagnose the issue more effectively. You can find more information on creating a reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).
> 
> Feel free to reach out if you have any more questions or need further assistance! 😊

There is still no attrubute called imgs

> AttributeError: 'Results' object has no attribute 'imgs'. See valid attributes below.

**Comment by glenn-jocher** [comment]
Thank you for catching that. To access the resized dimensions, you can use the `model.predict()` method and inspect the input size directly. If you need further assistance, please refer to the documentation or let us know.
